contracts/contracts/core/rlp-manager/src/main.sw

```diff
@@ -0,0 +1,663 @@
+        u64::try_from(mint_amount).unwrap()
+    );
+
+    log(AddLiquidity {
+        account,
+        asset,
+        amount,
+        aum_in_rusd,
+        rlp_supply,
+        rusd_amount,
+        mint_amount
+    });
+
+    mint_amount
+}
+
+#[storage(read, write)]
+fn _remove_liquidity(
+    account: Identity,
+    asset_out: AssetId,
+    rlp_amount: u64,
+    min_out: u64,
+    receiver: Identity
+) -> u256 {
+    require(
+        storage.last_added_at.get(account).try_read().unwrap_or(0) + storage.cooldown_duration.read() 
+            <= get_unix_timestamp(),
+        Error::RLPManagerCooldownDurationNotYetPassed
+    );
+
+    let vault = abi(Vault, storage.vault.read().into());
+
+    let rusd_contract = abi(YieldAsset, storage.rusd_contract.read().into());
+    let rusd_src3 = abi(SRC3, rusd_contract.get_id().into());
+    let rlp_contract = abi(RLP, storage.rlp_contract.read().into());
+    let rlp_src3 = abi(SRC3, storage.rlp_contract.read().into());
+    let rlp = rlp_contract.get_id();
+
+    require(
+        msg_asset_id() == rlp,
+        Error::RLPManagerInvalidRLPAssetForwarded
+    );
+    require(
+        msg_amount() == rlp_amount && msg_amount() > 0,
+        Error::RLPManagerInvalidRLPAmountForwarded
+    );
+
+    // calculate aum before sellRUSD
+    let aum_in_rusd = _get_aum_in_rusd(false);
+    let rlp_supply = rlp_contract.total_rlp_supply();
+
+    let rusd_amount = (rlp_amount.as_u256() * aum_in_rusd) / rlp_supply.as_u256();
+    let rusd_balance = balance_of(ContractId::this(), rusd_contract.get_id()).as_u256();
+    if rusd_amount > rusd_balance {
+        // @TODO: potential revert here
+        rusd_src3.mint(
+            Identity::ContractId(ContractId::this()), 
+            Some(DEFAULT_SUB_ID), // this is unused, but required by the interface to meet the SRC3 standard
+            u64::try_from(rusd_amount - rusd_balance).unwrap(),
+        );
+    }
+
+    // Burn RLP
+    rlp_src3.burn{
+        asset_id: rlp.into(),
+        coins: rlp_amount
+    }(
+        ZERO, // this is unused, but required by the interface to meet the SRC3 standard
+        rlp_amount
+    );
+
+    // sell RUSD
+    let amount_out = vault.sell_rusd{
+        asset_id: rusd_contract.get_id().into(),
+        // @TODO: potential revert here
+        coins: u64::try_from(rusd_amount).unwrap()
+    }(
+        asset_out,
+        receiver
+    );
+    require(
+        amount_out >= min_out.as_u256(), 
+        Error::RLPManagerInsufficientOutput
+    );
+
+    log(RemoveLiquidity {
+        account,
+        asset: asset_out,
+        rlp_amount,
+        aum_in_rusd,
+        rlp_supply,
+        rusd_amount,
+        amount_out
+    });
+
+    amount_out
+}
+
+#[storage(read)]
+fn _get_aum(maximize: bool) -> u256 {
+    let vault = abi(Vault, storage.vault.read().into());
+
+    let length = vault.get_all_whitelisted_assets_length();
+    let mut aum = storage.aum_addition.read();
```

Comment:
I don't understand this at all, we have aum_addition and subtration, but the subtration is never used, and why do we even need the aum_adjustment at all.

Reply:
rlp is removed